oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/NodeStateEntryWriter.java

```diff
@@ -118,14 +122,53 @@ public static String getPath(String entryLine) {
         return entryLine.substring(0, getDelimiterPosition(entryLine));
     }
 
+    /*
+        Currently we only have 2 types of store:
+        1. FlatFileStore
+        2. IncrementalFlatFileStore
+     */
     public static String[] getParts(String line) {
-        int pos = getDelimiterPosition(line);
-        return new String[]{line.substring(0, pos), line.substring(pos + 1)};
+        // file is FlatFileStore
+        if (line.endsWith("}")) {
+            int pos = getDelimiterPosition(line);
+            return new String[]{line.substring(0, pos), line.substring(pos + 1)};
+        } else {
+            // file is IncrementalFlatFileStore
+            List<Integer> positions = getDelimiterPositions(line);
+            checkState(positions.size() >= 2, "Invalid path entry [%s]", line);
+            // there are 4 parts in incrementalFFS and default delimiter is |
+            // path|nodeData|checkpoint|operand
+            // Node's data can itself have many | so we split based on first and last 2 |
+            String[] parts = new String[4];
+            parts[0] = line.substring(0, positions.get(0));
+            parts[1] = line.substring(positions.get(0) + 1, positions.get(positions.size() - 2));
+            parts[2] = line.substring(positions.get(positions.size() - 2) + 1, positions.get(positions.size() - 1));
+            parts[3] = line.substring(positions.get(positions.size() - 1) + 1);
+            return parts;
+        }
     }
 
     private static int getDelimiterPosition(String entryLine) {
         int indexOfPipe = entryLine.indexOf(NodeStateEntryWriter.DELIMITER);
         checkState(indexOfPipe > 0, "Invalid path entry [%s]", entryLine);
         return indexOfPipe;
     }
+
+    private static List<Integer> getDelimiterPositions(String entryLine) {
+        List<Integer> indexPositions = new LinkedList<>();
+        int index = 0;
+        int indexOfPipe;
+        while (true) {
+            indexOfPipe = entryLine.indexOf(NodeStateEntryWriter.DELIMITER, index);
+            if (indexOfPipe > 0) {
```

Comment:
Actually, it should be ">= 0"

Reply:
I think the JSON can contain the delimiter (NodeStateEntryWriter.DELIMITER). No?

The path can not. But the JSON, I think it can.

So we would have to get the first delimiter from the left. And the remaining 2 delimiter from the right (from end to beginning).